Add quantized adaptive avgpool.

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #36815 Add channel shuffle op fp32 + quantized.
* #36814 Replace empty_affine_quantizer with direct dispatch to at::native::empty_affine.. .
* **#36813 Add quantized adaptive avgpool.**
* #36812 Move to using MemoryFormat::ChannelsLast for quantized avgpool2d.

- Changes to q_avgpool to map special cases of adaptive avgpool to avgpool.
- Map special cases of adaptive avg pool to avgpool.

Differential Revision: [D21093837](https://our.internmc.facebook.com/intern/diff/D21093837/)

kimishpatel committed:
Add quantized adaptive avgpool.

- Changes to q_avgpool to map special cases of adaptive avgpool to avgpool.
- Map special cases of adaptive avg pool to avgpool.

Differential Revision: [D21093837](https://our.internmc.facebook.com/intern/diff/D21093837/)

[ghstack-poisoned]

z-a-f commented on test/quantization/test_quantized.py:
```diff
@@ -2789,6 +2789,54 @@ def test_avg_pool2d(
+            a_pool_q = torch.quantize_per_tensor(a_pool, scale=scale, zero_point=zero_point,
+                                                 dtype=torch.quint8)
+            np.testing.assert_array_almost_equal(a_pool_q.int_repr().numpy(),
+                                                 qa_pool.int_repr().numpy(), decimal=0)
```
Why is the `decimal=0`? Can we have a better reference? Alternatively, can we compare it to the results of the non-qnnpack quantized pool?

In `numpy` the `assert_array_almost_equal` with `decimal=0` is the same as `np.abs(difference) < 1.5`. In server side we use precision to be 1. I wonder if qnnpack numerics are different when comparing to the server-side kernel.

kimishpatel replied:
This is not deliberate, but we are not comparing floating point ones here. It is comparing quantized output.

z-a-f reviewed (approved):
LGTM, less a comment

dreiss commented on aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp:
```diff
@@ -227,11 +229,64 @@ Tensor q_adaptive_avg_pool2d(const Tensor& input, IntArrayRef output_size) {
+#ifdef USE_PYTORCH_QNNPACK
+Tensor qnnpack_adaptive_avg_pool2d(
+    const at::Tensor& input,
+    IntArrayRef output_size) {
+    std::array<int64_t, 2> kernel_size;
```
Does clang-format put these at the same indent?  Weird.

kimishpatel replied:
My bad. Fixing this. Thanks for catching.

dreiss commented on aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp:
```diff
@@ -227,11 +229,64 @@ Tensor q_adaptive_avg_pool2d(const Tensor& input, IntArrayRef output_size) {
+    auto input_height = input.sizes()[input.dim() - 2];
+    auto input_width = input.sizes()[input.dim() - 1];
+
+    return ((input_height % output_height == 0) &&
+        (input_width % output_width) == 0);
```
Is this true in practice?  I thought the point of adaptive average pool was that you wanted to shrink a variable-sized input into a fixed-sized output.

kimishpatel replied:
So this helps us reduce this to regular avgpool. If this condition is not met we fallback to the other slow path from here, https://github.com/pytorch/pytorch/pull/36813/files#diff-c227aae55653cf48955fdf36f92378afR286.
This is because qnnpack does not have adaptive avg pool implemented. I plan to implement this in PT qnnpack.

dreiss replied:
I understand that this is safe, but my question is how often this is true?  If this is an uncommon case, I don't think it's worth having an optimized implementation for it.

kimishpatel replied:
I dont know the extent to which this is common but this is for our internal mobile vision models which I was trying to optimize. Thats what they have. And the unoptimized path is much slower.
Maybe we can remove this once I have qnnpack implementation for the adaptive avgpool but that will still be slower in degenerate cases like this.

dreiss commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/average-pooling.c:
```diff
@@ -270,6 +270,67 @@ enum pytorch_qnnp_status pytorch_qnnp_setup_average_pooling2d_nhwc_q8(
+   *  | | | | | | | |
+   *  ---------------
+   *
+   *  Thus we are going from width=7 height=5 input to height=2 width=3
```
Why did you switch from width,height to height,width?  Is that an error?

kimishpatel replied:
Thanks for the catch. I will make it consistent.

kimishpatel committed:
Update on "Add quantized adaptive avgpool."


- Changes to q_avgpool to map special cases of adaptive avgpool to avgpool.
- Map special cases of adaptive avg pool to avgpool.

Differential Revision: [D21093837](https://our.internmc.facebook.com/intern/diff/D21093837/)

[ghstack-poisoned]

kimishpatel committed:
Update on "Add quantized adaptive avgpool."


- Changes to q_avgpool to map special cases of adaptive avgpool to avgpool.
- Map special cases of adaptive avg pool to avgpool.

Differential Revision: [D21093837](https://our.internmc.facebook.com/intern/diff/D21093837/)

[ghstack-poisoned]

kimishpatel committed:
Update on "Add quantized adaptive avgpool."


- Changes to q_avgpool to map special cases of adaptive avgpool to avgpool.
- Map special cases of adaptive avg pool to avgpool.

Differential Revision: [D21093837](https://our.internmc.facebook.com/intern/diff/D21093837/)

[ghstack-poisoned]

kimishpatel committed:
Update on "Add quantized adaptive avgpool."


- Changes to q_avgpool to map special cases of adaptive avgpool to avgpool.
- Map special cases of adaptive avg pool to avgpool.

Differential Revision: [D21093837](https://our.internmc.facebook.com/intern/diff/D21093837/)

[ghstack-poisoned]

kimishpatel committed:
Update on "Add quantized adaptive avgpool."


- Changes to q_avgpool to map special cases of adaptive avgpool to avgpool.
- Map special cases of adaptive avg pool to avgpool.

Differential Revision: [D21093837](https://our.internmc.facebook.com/intern/diff/D21093837/)

[ghstack-poisoned]

kimishpatel committed:
Update on "Add quantized adaptive avgpool."


- Changes to q_avgpool to map special cases of adaptive avgpool to avgpool.
- Map special cases of adaptive avg pool to avgpool.

Differential Revision: [D21093837](https://our.internmc.facebook.com/intern/diff/D21093837/)

[ghstack-poisoned]
